src/contracts/core/AVSDirectory.sol

```diff
@@ -249,114 +265,192 @@ contract AVSDirectory is
+    function slashOperator(
         address operator,
-        ISignatureUtils.SignatureWithSaltAndExpiry memory operatorSignature
-    ) external override onlyWhenNotPaused(PAUSED_OPERATOR_REGISTER_DEREGISTER_TO_AVS) {
-        // Assert `operatorSignature.expiry` has not elapsed.
+        uint32 operatorSetId,
+        IStrategy[] calldata strategies,
+        uint16 bipsToSlash
+    ) external {
         require(
-            operatorSignature.expiry >= block.timestamp,
-            "AVSDirectory.registerOperatorToAVS: operator signature expired"
+            isOperatorSlashable(operator, OperatorSet({avs: msg.sender, operatorSetId: operatorSetId})),
+            "AVSDirectory.slashOperator: operator not slashable for operatorSet"
         );
 
-        // Assert that the AVS is not an operator set AVS.
-        require(!isOperatorSetAVS[msg.sender], "AVSDirectory.registerOperatorToAVS: AVS is an operator set AVS");
+        for (uint256 i = 0; i < strategies.length; ++i) {
+            // 1. calculate slashed magnitude from current allocation
+            // update current and all following queued magnitude updates for (operator, strategy, operatorSetId) tuple
+            uint64 slashedMagnitude;
+            {
+                uint64 currentMagnitude = uint64(
+                    _magnitudeUpdate[operator][strategies[i]][msg.sender][operatorSetId].upperLookupRecent(
+                        uint32(block.timestamp)
+                    )
+                );
+                // TODO: if we don't continue here we get into weird "total/free magnitude" not initialized cases. Is this ok?
+                if (currentMagnitude == 0) {
+                    continue;
+                }
+
+                /// TODO: add wrapping library for rounding up for slashing accounting
+                slashedMagnitude = uint64(uint256(bipsToSlash) * uint256(currentMagnitude) / BIPS_FACTOR);
+
+                _magnitudeUpdate[operator][strategies[i]][msg.sender][operatorSetId].decrementAtAndFutureCheckpoints({
+                    key: uint32(block.timestamp),
+                    decrementValue: slashedMagnitude
+                });
+            }
+
+            // 2. if there are any pending deallocations then need to update and decrement if they fall within slashable window
+            // loop backwards through _queuedDeallocationIndices, each element contains an array index to
+            // corresponding deallocation to access in pendingFreeMagnitude
+            // if completable, then break
+            //      (since ordered by completableTimestamps, older deallocations will also be completable and outside slashable window)
+            // if NOT completable, then slash
+            {
+                uint256 queuedDeallocationIndicesLen =
+                    _queuedDeallocationIndices[operator][strategies[i]][msg.sender][operatorSetId].length;
+                for (uint256 j = queuedDeallocationIndicesLen; j > 0; --j) {
```

Comment:
Since `MAX_PENDING_UPDATES` is 1, we don't need a loop here?

Reply:
@gpsanant mentioned we want to check without assuming max 1 and be able to update in the future to change `MAX_PENDING_UPDATES`

Reply:
eh

Reply:
@gpsanant Should I remove the loop?